gamestonk_terminal/loggers.py

```diff
@@ -0,0 +1,181 @@
+    handler.setFormatter(formatter)
+    logging.getLogger().addHandler(handler)
+
+
+class CustomFormatterWithExceptions(logging.Formatter):
+    """Custom Logging Formatter that includes formatting of Exceptions"""
+
+    def formatException(self, ei) -> str:
+        """Exception formatting handler
+
+        Parameters
+        ----------
+        ei : logging._SysExcInfoType
+            Exception to be logged
+
+        Returns
+        -------
+        str
+            Formatted exception
+        """
+        result = super().formatException(ei)
+        return repr(result)
+
+    def format(self, record: logging.LogRecord) -> str:
+        """Log formatter
+
+        Parameters
+        ----------
+        record : logging.LogRecord
+            Logging record
+
+        Returns
+        -------
+        str
+            Formatted log message
+        """
+        s = super().format(record)
+        if record.levelname:
+            prefix = record.levelname[0]
+        else:
+            prefix = "U"
+
+        if record.exc_text:
+            s = "X|" + cfg.LOGGING_VERSION + "|" + s.replace("\n", "|") + "|"
+        else:
+            s = f"{prefix}|{cfg.LOGGING_VERSION}|{s}|none|"
+
+        return s
+
+
+def get_commit_hash() -> None:
+    """Get Commit Short Hash"""
+
+    file_path = Path(__file__)
+    git_dir = file_path.parent.parent.absolute().joinpath(".git")
+
+    if os.path.isdir(git_dir.absolute()):
+        repo = git.Repo(search_parent_directories=True)
+        sha = repo.head.object.hexsha
+        short_sha = repo.git.rev_parse(sha, short=8)
+        cfg.LOGGING_VERSION = f"sha:{short_sha}"
+
+
+def setup_logging() -> None:
+    """Setup Logging"""
+
+    if int(cfg.LOGGING_VERBOSITY) < 1:
+        verbosity = logging.INFO
+    else:
+        verbosity = logging.DEBUG
+
+    logging.basicConfig(
+        level=verbosity, format=LOGFORMAT, datefmt=DATEFORMAT, handlers=[]
+    )
+
+    get_commit_hash()
+
+    for a_handler in cfg.LOGGING_HANDLERS.split(","):
+        if a_handler == "stdout":
+            handler = logging.StreamHandler(sys.stdout)
+            formatter = CustomFormatterWithExceptions(fmt=LOGFORMAT, datefmt=DATEFORMAT)
+            handler.setFormatter(formatter)
+            logging.getLogger().addHandler(handler)
+        elif a_handler == "stderr":
+            handler = logging.StreamHandler(sys.stderr)
+            formatter = CustomFormatterWithExceptions(fmt=LOGFORMAT, datefmt=DATEFORMAT)
+            handler.setFormatter(formatter)
+            logging.getLogger().addHandler(handler)
+        elif a_handler == "noop":
+            handler = logging.NullHandler()  # type: ignore
+            formatter = CustomFormatterWithExceptions(fmt=LOGFORMAT, datefmt=DATEFORMAT)
+            handler.setFormatter(formatter)
+            logging.getLogger().addHandler(handler)
+        elif a_handler == "file":
+            setup_file_logger()
+        else:
+            logger.debug("Unknown loghandler")
```

Comment:
To avoid code repetition we can have
```
for a_handler in cfg.LOGGING_HANDLERS.split(","):
         if a_handler == "stdout":
             handler = logging.StreamHandler(sys.stdout)
         elif a_handler == "stderr":
             handler = logging.StreamHandler(sys.stderr)
         elif a_handler == "noop":
             handler = logging.NullHandler()  # type: ignore
         elif a_handler == "file":
             setup_file_logger()
         else:
             logger.debug("Unknown loghandler")

        if a_handler in ("stdout", "stderr", "noop"):
             formatter = CustomFormatterWithExceptions(fmt=LOGFORMAT, datefmt=DATEFORMAT)
             handler.setFormatter(formatter)
             logging.getLogger().addHandler(handler)
```

Reply:
That looks less clear to me.